lightning/src/util/base32.rs

```diff
@@ -0,0 +1,281 @@
+// This is a modification of base32 encoding to support the zbase32 alphabet.
+// The original piece of software can be found at https://crates.io/crates/base32(v0.4.0)
+// The original portions of this software are Copyright (c) 2015 The base32 Developers
+
+// This file is licensed under either of
+// Apache License, Version 2.0, (LICENSE-APACHE or http://www.apache.org/licenses/LICENSE-2.0) or
+// MIT license (LICENSE-MIT or http://opensource.org/licenses/MIT) at your option.
+
+
+use crate::prelude::*;
+
+/// RFC4648 encoding table
+const RFC4648_ALPHABET: &'static [u8] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";
+
+/// Zbase encoding alphabet
+const ZBASE_ALPHABET: &'static [u8] = b"ybndrfg8ejkmcpqxot1uwisza345h769";
+
+/// RFC4648 decoding table
+const RFC4648_INV_ALPHABET: [i8; 43] = [
+	-1, -1, 26, 27, 28, 29, 30, 31, -1, -1, -1, -1, -1, 0, -1, -1, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8,
+	9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25,
+];
+
+/// Zbase decoding table
+const ZBASE_INV_ALPHABET: [i8; 43] = [
+	-1, 18, -1, 25, 26, 27, 30, 29, 7, 31, -1, -1, -1, -1, -1, -1, -1, 24, 1, 12, 3, 8, 5, 6, 28,
+	21, 9, 10, -1, 11, 2, 16, 13, 14, 4, 22, 17, 19, -1, 20, 15, 0, 23,
+];
+
+/// Alphabet used for encoding and decoding.
+#[derive(Copy, Clone)]
+pub enum Alphabet {
+	/// RFC4648 encoding.
+	RFC4648 {
+		/// Whether to use padding.
+		padding: bool
+	},
+	/// Zbase32 encoding.
+	ZBase32
+}
+
+impl Alphabet {
+	/// Encode bytes into a base32 string.
+	pub fn encode(&self, data: &[u8]) -> Result<String, ()> {
+		let ret = match self {
+			Self::RFC4648 { padding } =>{
+				let mut ret = Self::encode_data(data, RFC4648_ALPHABET);
+				if data.len() % 5 != 0 {
+					let len = ret.len();
+					let num_extra = 8 - (data.len() % 5 * 8 + 4) / 5;
+					if *padding {
+						for i in 1..num_extra + 1 {
+							ret[len - i] = b'=';
+						}
+					} else {
+						ret.truncate(len - num_extra);
+					}
+				}
+				ret
+
+			},
+			Self::ZBase32 => {
+				let mut ret = Self::encode_data(data, ZBASE_ALPHABET);
+				ret.truncate((data.len() * 8 + 4) / 5);
+				ret
+			}
+		};
+		#[cfg(fuzzing)]
+		assert_eq!(ret.capacity(), (data.len() + 4) / 5 * 8);
+
+		String::from_utf8(ret).map_err(|_| ())
+	}
+
+	/// Decode a base32 string into a byte vector.
+	pub fn decode(&self, data: &str) -> Result<Vec<u8>, ()> {
+		let data = data.as_bytes();
+		let ret = match self {
+			Self::RFC4648 { .. } => {
+				if data.len() % 8 != 0 {
+					return Err(());
+				}
+				let mut ret = Self::decode_data(data, RFC4648_INV_ALPHABET)?;
+				let mut unpadded_data_length = data.len();
+				data.iter().rev().take(6).for_each(|&c| {
+					if c == b'=' {
+						unpadded_data_length -= 1;
+					}
+				});
+				let output_length = unpadded_data_length * 5 / 8;
+				ret.truncate(output_length);
+				ret
+
+			},
+			Self::ZBase32 => {
+				let output_length = data.len() * 5 / 8;
+				if data.len() > (output_length * 8 + 4) / 5 {
```

Comment:
why not? I think its valid because we don't have padding in zbase
the fuzzer also wouldnt be happy if we remove it, would fail with "ZrY" for example